doseCalc/+DoseEngines/matRad_PencilBeamEngineAbstract.m

```diff
@@ -0,0 +1,643 @@
+
+                                    %First, create a ray of the
+                                    %specific scenario to adapt rad
+                                    %depths
+                                    scenRay = currRay;
+                                    scenRay.radDepths = scenRay.radDepths{ctScen};
+                                    scenRay.radDepths = (1+this.multScen.relRangeShift(scenNum))*scenRay.radDepths + this.multScen.absRangeShift(scenNum);
+
+                                    for k = 1:currRay.numOfBixels
+                                        
+                                                                                
+                                        %Bixel Computation
+                                        currBixel = this.computeBixel(scenRay,k);
+
+                                        % save computation time and memory
+                                        % by sequentially filling the sparse matrix dose.dij from the cell array
+                                        dij = this.fillDij(currBixel,dij,scenStf,scenIdx,i,j,k,bixelCounter + k);
+                                    end
+                                end
+                            end
+                        end
+
+                        % Progress Update & Bookkeeping
+                        bixelCounter = bixelCounter + currRay.numOfBixels;
+                        bixelBeamCounter = bixelBeamCounter + currRay.numOfBixels;
+                        this.progressUpdate(bixelCounter,dij.totalNumOfBixels);
+                    end
+                end
+            end
+
+            %Finalize dose calculation
+            dij = this.finalizeDose(dij);
+        end
+
+        function dij = initDoseCalc(this,ct,cst,stf)
+            % modified inherited method of the superclass DoseEngine,
+            % containing intialization which are specificly needed for
+            % pencil beam calculation and not for other engines
+
+            dij = initDoseCalc@DoseEngines.matRad_DoseEngineBase(this,ct,cst,stf);
+            
+            % calculate rED or rSP from HU or take provided wedCube
+            if this.useGivenEqDensityCube && ~isfield(ct,'cube')
+                matRad_cfg.dispWarning('HU Conversion requested to be omitted but no ct.cube exists! Will override and do the conversion anyway!');
+                this.useGivenEqDensityCube = false;
+            end
+
+            if this.useGivenEqDensityCube
+                matRad_cfg.dispInfo('Omitting HU to rED/rSP conversion and using existing ct.cube!\n');
+            else
+                ct = matRad_calcWaterEqD(ct, stf); % Maybe we can avoid duplicating the CT here?
+            end
+
+            this.cubeWED = ct.cube;
+            if isfield(ct,'hlut')
+                this.hlut = ct.hlut;
+            end
+
+            % ignore densities outside of contours
+            if this.ignoreOutsideDensities
```

Comment:
Here initDoseCalc modifies the ct cube but does not output the modified structure. I think changes applied to `ct` here are forgotten

Reply:
It now applies the changes to `this.cubeWED` which will be used later for raytracing, so changes are persistent now.